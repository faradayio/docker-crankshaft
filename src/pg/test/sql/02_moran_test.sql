\pset format unaligned
\set ECHO all
\i test/fixtures/ppoints.sql
\i test/fixtures/ppoints2.sql

-- Areas of Interest functions perform some nondeterministic computations
-- (to estimate the significance); we will set the seeds for the RNGs
-- that affect those results to have repeateble results

-- Moran's I Global
SELECT cdb_crankshaft._cdb_random_seeds(1234);

SELECT round(moran, 4) As moran, round(significance, 4) As significance
  FROM cdb_crankshaft.CDB_AreasOfInterestGlobal('SELECT * FROM ppoints', 'value') m(moran, significance);

-- Moran's I Local
SELECT cdb_crankshaft._cdb_random_seeds(1234);

SELECT ppoints.code, m.quads
  FROM ppoints
  JOIN cdb_crankshaft.CDB_AreasOfInterestLocal('SELECT * FROM ppoints', 'value') m
    ON ppoints.cartodb_id = m.rowid
  ORDER BY ppoints.code;

SELECT cdb_crankshaft._cdb_random_seeds(1234);

-- Moran's I local
SELECT
  ppoints.code, m.quads,
  abs(avg(m.orig_val_std) OVER ()) < 1e-6 as diff_orig,
  CASE WHEN m.quads = 'HL' THEN m.orig_val_std > m.spatial_lag_std
       WHEN m.quads = 'HH' THEN m.orig_val_std >= 0 and m.spatial_lag_std >= 0
       WHEN m.quads = 'LH' THEN m.orig_val_std < m.spatial_lag_std
       WHEN m.quads = 'LL' THEN m.orig_val_std <= 0 and m.spatial_lag_std <= 0
       ELSE null END as expected,
  moran_stat is not null moran_stat_not_null,
  significance >= 0.001 significance_not_null, -- greater than 1/1000 (default)
  abs(m.orig_val - ppoints.value) <= 1e-6 as value_comparison
  FROM ppoints
  JOIN cdb_crankshaft.CDB_MoransILocal('SELECT * FROM ppoints', 'value') m
    ON ppoints.cartodb_id = m.rowid
  ORDER BY ppoints.code;

SELECT cdb_crankshaft._cdb_random_seeds(1234);

-- Spatial Hotspots
SELECT ppoints.code, m.quads
  FROM ppoints
  JOIN cdb_crankshaft.CDB_GetSpatialHotspots('SELECT * FROM ppoints', 'value') m
    ON ppoints.cartodb_id = m.rowid
  ORDER BY ppoints.code;

SELECT cdb_crankshaft._cdb_random_seeds(1234);

-- Spatial Coldspots
SELECT ppoints.code, m.quads
  FROM ppoints
  JOIN cdb_crankshaft.CDB_GetSpatialColdspots('SELECT * FROM ppoints', 'value') m
    ON ppoints.cartodb_id = m.rowid
  ORDER BY ppoints.code;

SELECT cdb_crankshaft._cdb_random_seeds(1234);

  -- Spatial Outliers
SELECT ppoints.code, m.quads
  FROM ppoints
  JOIN cdb_crankshaft.CDB_GetSpatialOutliers('SELECT * FROM ppoints', 'value') m
    ON ppoints.cartodb_id = m.rowid
  ORDER BY ppoints.code;


SELECT cdb_crankshaft._cdb_random_seeds(1234);

-- Areas of Interest (rate)
SELECT ppoints2.code, m.quads
  FROM ppoints2
  JOIN cdb_crankshaft.CDB_AreasOfInterestLocalRate('SELECT * FROM ppoints2', 'numerator', 'denominator') m
    ON ppoints2.cartodb_id = m.rowid
  ORDER BY ppoints2.code;

SELECT cdb_crankshaft._cdb_random_seeds(1234);

-- Moran's I local rate
SELECT
  ppoints2.code, m.quads,
  abs(avg(m.orig_val_std) OVER ()) < 1e-6 as diff_orig,
  CASE WHEN m.quads = 'HL' THEN m.orig_val_std > m.spatial_lag_std
       WHEN m.quads = 'HH' THEN m.orig_val_std >= 0 and m.spatial_lag_std >= 0
       WHEN m.quads = 'LH' THEN m.orig_val_std < m.spatial_lag_std
       WHEN m.quads = 'LL' THEN m.orig_val_std <= 0 and m.spatial_lag_std <= 0
       ELSE null END as expected,
  moran_stat is not null moran_stat_not_null,
  significance >= 0.001 significance_not_null -- greater than 1/1000 (default)
  FROM ppoints2
  JOIN cdb_crankshaft.CDB_MoransILocalRate('SELECT * FROM ppoints2', 'numerator', 'denominator') m
    ON ppoints2.cartodb_id = m.rowid
  ORDER BY ppoints2.code;

SELECT cdb_crankshaft._cdb_random_seeds(1234);

-- Spatial Hotspots (rate)
SELECT ppoints2.code, m.quads
  FROM ppoints2
  JOIN cdb_crankshaft.CDB_GetSpatialHotspotsRate('SELECT * FROM ppoints2', 'numerator', 'denominator') m
    ON ppoints2.cartodb_id = m.rowid
  ORDER BY ppoints2.code;

SELECT cdb_crankshaft._cdb_random_seeds(1234);

-- Spatial Coldspots (rate)
SELECT ppoints2.code, m.quads
  FROM ppoints2
  JOIN cdb_crankshaft.CDB_GetSpatialColdspotsRate('SELECT * FROM ppoints2', 'numerator', 'denominator') m
    ON ppoints2.cartodb_id = m.rowid
  ORDER BY ppoints2.code;

SELECT cdb_crankshaft._cdb_random_seeds(1234);

-- Spatial Outliers (rate)
SELECT ppoints2.code, m.quads
  FROM ppoints2
  JOIN cdb_crankshaft.CDB_GetSpatialOutliersRate('SELECT * FROM ppoints2', 'numerator', 'denominator') m
    ON ppoints2.cartodb_id = m.rowid
  ORDER BY ppoints2.code;
