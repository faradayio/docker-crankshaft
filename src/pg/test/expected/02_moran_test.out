\pset format unaligned
\set ECHO all
\i test/fixtures/ppoints.sql
SET client_min_messages TO WARNING;
\set ECHO none
_cdb_random_seeds

(1 row)
moran|significance
0.3399|-0.0196
(1 row)
_cdb_random_seeds

(1 row)
code|quads
01|HH
02|HL
03|LL
04|LL
05|LH
06|LL
07|HH
08|HH
09|HH
10|LL
11|LL
12|LL
13|HL
14|LL
15|LL
16|HH
17|HH
18|LL
19|HH
20|HH
21|LL
22|HH
23|LL
24|LL
25|HH
26|HH
27|LL
28|HH
29|LL
30|LL
31|HH
32|LL
33|HL
34|LH
35|LL
36|LL
37|HL
38|HL
39|HH
40|HH
41|HL
42|LH
43|LH
44|LL
45|LH
46|LL
47|LL
48|HH
49|LH
50|HH
51|LL
52|LL
(52 rows)
_cdb_random_seeds

(1 row)
code|quads|diff_orig|expected|moran_stat_not_null|significance_not_null|value_comparison
01|HH|t|t|t|t|t
02|HL|t|t|t|t|t
03|LL|t|t|t|t|t
04|LL|t|t|t|t|t
05|LH|t|t|t|t|t
06|LL|t|t|t|t|t
07|HH|t|t|t|t|t
08|HH|t|t|t|t|t
09|HH|t|t|t|t|t
10|LL|t|t|t|t|t
11|LL|t|t|t|t|t
12|LL|t|t|t|t|t
13|HL|t|t|t|t|t
14|LL|t|t|t|t|t
15|LL|t|t|t|t|t
16|HH|t|t|t|t|t
17|HH|t|t|t|t|t
18|LL|t|t|t|t|t
19|HH|t|t|t|t|t
20|HH|t|t|t|t|t
21|LL|t|t|t|t|t
22|HH|t|t|t|t|t
23|LL|t|t|t|t|t
24|LL|t|t|t|t|t
25|HH|t|t|t|t|t
26|HH|t|t|t|t|t
27|LL|t|t|t|t|t
28|HH|t|t|t|t|t
29|LL|t|t|t|t|t
30|LL|t|t|t|t|t
31|HH|t|t|t|t|t
32|LL|t|t|t|t|t
33|HL|t|t|t|t|t
34|LH|t|t|t|t|t
35|LL|t|t|t|t|t
36|LL|t|t|t|t|t
37|HL|t|t|t|t|t
38|HL|t|t|t|t|t
39|HH|t|t|t|t|t
40|HH|t|t|t|t|t
41|HL|t|t|t|t|t
42|LH|t|t|t|t|t
43|LH|t|t|t|t|t
44|LL|t|t|t|t|t
45|LH|t|t|t|t|t
46|LL|t|t|t|t|t
47|LL|t|t|t|t|t
48|HH|t|t|t|t|t
49|LH|t|t|t|t|t
50|HH|t|t|t|t|t
51|LL|t|t|t|t|t
52|LL|t|t|t|t|t
(52 rows)
_cdb_random_seeds

(1 row)
code|quads
01|HH
02|HL
07|HH
08|HH
09|HH
13|HL
16|HH
17|HH
19|HH
20|HH
22|HH
25|HH
26|HH
28|HH
31|HH
33|HL
37|HL
38|HL
39|HH
40|HH
41|HL
48|HH
50|HH
(23 rows)
_cdb_random_seeds

(1 row)
code|quads
03|LL
04|LL
05|LH
06|LL
10|LL
11|LL
12|LL
14|LL
15|LL
18|LL
21|LL
23|LL
24|LL
27|LL
29|LL
30|LL
32|LL
34|LH
35|LL
36|LL
42|LH
43|LH
44|LL
45|LH
46|LL
47|LL
49|LH
51|LL
52|LL
(29 rows)
_cdb_random_seeds

(1 row)
code|quads
02|HL
05|LH
13|HL
33|HL
34|LH
37|HL
38|HL
41|HL
42|LH
43|LH
45|LH
49|LH
(12 rows)
_cdb_random_seeds

(1 row)
code|quads
01|HH
02|HL
03|LL
04|LL
05|LH
06|LL
07|HH
08|HH
09|HH
10|LL
11|LL
12|LL
13|HL
14|LL
15|LL
16|HH
17|HH
18|LL
19|HH
20|HH
21|LL
22|HH
23|LL
24|LL
25|HH
26|HH
27|LL
28|HH
29|LL
30|LL
31|HH
32|LL
33|HL
34|LH
35|LL
36|LL
37|HL
38|HL
39|HH
40|HH
41|HL
42|LH
43|LH
44|LL
45|LH
46|LL
47|LL
48|HH
49|LH
50|HH
51|LL
52|LL
(52 rows)
_cdb_random_seeds

(1 row)
code|quads|diff_orig|expected|moran_stat_not_null|significance_not_null
01|HH|t|t|t|t
02|HL|t|t|t|t
03|LL|t|t|t|t
04|LL|t|t|t|t
05|LH|t|t|t|t
06|LL|t|t|t|t
07|HH|t|t|t|t
08|HH|t|t|t|t
09|HH|t|t|t|t
10|LL|t|t|t|t
11|LL|t|t|t|t
12|LL|t|t|t|t
13|HL|t|t|t|t
14|LL|t|t|t|t
15|LL|t|t|t|t
16|HH|t|t|t|t
17|HH|t|t|t|t
18|LL|t|t|t|t
19|HH|t|t|t|t
20|HH|t|t|t|t
21|LL|t|t|t|t
22|HH|t|t|t|t
23|LL|t|t|t|t
24|LL|t|t|t|t
25|HH|t|t|t|t
26|HH|t|t|t|t
27|LL|t|t|t|t
28|HH|t|t|t|t
29|LL|t|t|t|t
30|LL|t|t|t|t
31|HH|t|t|t|t
32|LL|t|t|t|t
33|HL|t|t|t|t
34|LH|t|t|t|t
35|LL|t|t|t|t
36|LL|t|t|t|t
37|HL|t|t|t|t
38|HL|t|t|t|t
39|HH|t|t|t|t
40|HH|t|t|t|t
41|HL|t|t|t|t
42|LH|t|t|t|t
43|LH|t|t|t|t
44|LL|t|t|t|t
45|LH|t|t|t|t
46|LL|t|t|t|t
47|LL|t|t|t|t
48|HH|t|t|t|t
49|LH|t|t|t|t
50|HH|t|t|t|t
51|LL|t|t|t|t
52|LL|t|t|t|t
(52 rows)
_cdb_random_seeds

(1 row)
code|quads
01|HH
02|HL
07|HH
08|HH
09|HH
13|HL
16|HH
17|HH
19|HH
20|HH
22|HH
25|HH
26|HH
28|HH
31|HH
33|HL
37|HL
38|HL
39|HH
40|HH
41|HL
48|HH
50|HH
(23 rows)
_cdb_random_seeds

(1 row)
code|quads
03|LL
04|LL
05|LH
06|LL
10|LL
11|LL
12|LL
14|LL
15|LL
18|LL
21|LL
23|LL
24|LL
27|LL
29|LL
30|LL
32|LL
34|LH
35|LL
36|LL
42|LH
43|LH
44|LL
45|LH
46|LL
47|LL
49|LH
51|LL
52|LL
(29 rows)
_cdb_random_seeds

(1 row)
code|quads
02|HL
05|LH
13|HL
33|HL
34|LH
37|HL
38|HL
41|HL
42|LH
43|LH
45|LH
49|LH
(12 rows)
